src/main/java/org/opensearch/geospatial/ip2geo/action/PutDatasourceTransportAction.java

```diff
@@ -52,138 +41,96 @@
-        clusterSettings.addSettingsUpdateConsumer(Ip2GeoSettings.INDEXING_BULK_SIZE, newValue -> indexingBulkSize = newValue);
+        this.datasourceFacade = datasourceFacade;
+        this.datasourceUpdateService = datasourceUpdateService;
     }
 
     @Override
     protected void doExecute(final Task task, final PutDatasourceRequest request, final ActionListener<AcknowledgedResponse> listener) {
         try {
-            Datasource jobParameter = Datasource.Builder.build(request);
+            Datasource datasource = Datasource.Builder.build(request);
             IndexRequest indexRequest = new IndexRequest().index(DatasourceExtension.JOB_INDEX_NAME)
-                .id(jobParameter.getId())
-                .source(jobParameter.toXContent(JsonXContent.contentBuilder(), null))
+                .id(datasource.getId())
+                .source(datasource.toXContent(JsonXContent.contentBuilder(), null))
                 .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
                 .opType(DocWriteRequest.OpType.CREATE);
-            client.index(indexRequest, new ActionListener<>() {
-                @Override
-                public void onResponse(final IndexResponse indexResponse) {
-                    // This is user initiated request. Therefore, we want to handle the first datasource update task in a generic thread
-                    // pool.
-                    threadPool.generic().submit(() -> {
-                        try {
-                            createDatasource(jobParameter);
-                        } catch (Exception e) {
-                            log.error("Failed to create datasource for {}", jobParameter.getId(), e);
-                            jobParameter.getUpdateStats().setLastFailedAt(Instant.now());
-                            jobParameter.setState(DatasourceState.FAILED);
-                            try {
-                                DatasourceHelper.updateDatasource(client, jobParameter, timeout);
-                            } catch (Exception ex) {
-                                log.error("Failed to mark datasource state as FAILED for {}", jobParameter.getId(), ex);
-                            }
-                        }
-                    });
-                    listener.onResponse(new AcknowledgedResponse(true));
-                }
-
-                @Override
-                public void onFailure(final Exception e) {
-                    if (e instanceof VersionConflictEngineException) {
-                        listener.onFailure(
-                            new ResourceAlreadyExistsException("datasource [{}] already exists", request.getDatasourceName())
-                        );
-                    } else {
-                        listener.onFailure(e);
-                    }
-                }
-            });
+            client.index(indexRequest, getIndexResponseListener(datasource, listener));
```

Comment:
Will this implicitly create the JOB_INDEX_NAME index? Do we want to explicitly control the creation of the index so that we explicitly set mapping as well as other settings?

Reply:
This index is controlled by job scheduler framework. They create the index accordingly. 
See. https://github.com/opensearch-project/geospatial/blob/feature/ip2geo/src/main/java/org/opensearch/geospatial/ip2geo/jobscheduler/DatasourceExtension.java#L34

Reply:
Above is our plugin. But I think we might need to create the index itself. In index management, they create the index on their own: https://github.com/opensearch-project/index-management/blob/2.7/src/main/kotlin/org/opensearch/indexmanagement/IndexManagementIndices.kt#L60

Reply:
I thought job scheduler will create index for us but seems it is not. Let me create the index explicitly in a separate PR as current PR is already too big to add another change here.